### ISSUE TITLE
HlsChunkSource starts with wrong track, despite resolution/bitrate constraints

### ISSUE THREAD

**Issue by davibe** [google/ExoPlayer#3079]
### Issue description
As a user/integrator of exoplayer i want to be able to set max bandwidth and/or resolution for the exoplayer HLS adaptation logic. I would expect the player to never download and try to playback segments that do not satisfy the constraints I have set.
However, when the HLS manifest is written with inner playlists sorted from the highest to the lowest quality ExoPlayer will try to start with the highest quality no matter what the constraints are.
Devices like Samsung S4 mini will just crash right away when trying to decode segments with very high resolutions.

### Link to test content
http://dmvideo.deltatre.it/VR/TEST_MBR/main.m3u

### Version of ExoPlayer being used
Latest exoplayer release-v2 branch as of today

### Device(s) and version(s) of Android being used
Samsung S4 mini is one of the device which needs to have resolution/bitrate constraints. Android 4.4.2

### A full bug report captured from the device
This is the output taken right after the Samsung S4 mini fails to play the manifest file as described above.
[bugreport-1.txt](https://github.com/google/ExoPlayer/files/1156251/bugreport-1.txt)

### Further discussion after hours of debugging

Ref.
https://github.com/google/ExoPlayer/blob/release-v2/library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java#L397

`group` contains a list of formats ordered as they are ordered in the manifest file.
`indexOf()` instead looks up the given index in the list of formats sorted by decreasing bandwidth.

Using `group.getFormat(0)` may return either the lowest or the highest available bitrate/resolution depending on how the manifest was written. This effectively means that exoplayer MAY start loading segment starting with the highest or the lowest quality chunk before actual adaptation logic kicks in.

On some devices like the Samsung S4 mini we need to put constraints (maxBitrate, maxVideoHeight, maxVideoWidth) otherwise the device decoder will fail. Because of the aforementioned problem our constraints have no effects. Even with constraints the playback may still start from an unpredictable bitrate/resolution depending on the manifest and the device fails to decode (no playback).

**Comment by ojw28** [comment]
I'm really struggling to understand this issue. Primarily due to the lack of a bug report, as is explicitly requested in the issue template. If you take a bug report shortly after playing your test stream using the demo app, it'll include detailed debugging information showing us what actually happened during the playback, including the track selections and adaptations that took place. Please play the sample in the demo app and capture a bug report.

Regarding this:

> Using group.getFormat(0) may return either the lowest or the highest available bitrate/resolution depending on how the manifest was written. This effectively means that exoplayer MAY start loading segment starting with the highest or the lowest quality chunk before actual adaptation logic kicks in.

It doesn't (or shouldn't) matter, since we'll discard and re-buffer a different format if the track selector indicates that a different stream should be selected after preparation completes. None of the media from the first format will be fed to the device decoders if this occurs.

> Devices like Samsung S4 mini will just crash right away when trying to decode segments with very high resolutions.

Our track selection logic does automatically avoid selecting streams that the device indicates it cannot support. So in general it shouldn't be necessary for you to do this kind of thing. It is possible that certain devices are over-reporting their own capabilities, however.

**Comment by davibe** [comment]
I know a proper bug report would be better but I am not able to provide a publicly available example manifest at the moment.

Despite what you said I can guarantee that with a manifest file as the one described above (sorted from the highest to the lowest quality) the first segments downloaded are the ones with very high resolution and they are used to initialize the decoder resulting in a failure.
This is true using the official exoplayer demo just feeding it the url of that manifest.

I have asked my company if they can provide a public example and if so I will update this. In the mean time this is how I fixed it: https://github.com/davibe/ExoPlayer/commit/39e2980666b15f465133e4aee1bf84a3659dd26a
Basically with this commit It always starts with the lowest quality fragment because it looks up the last one in the set of formats sorted by quality desc.

**Comment by ojw28** [comment]
You do not need to provide a publicly available manifest to provide a bug report captured as described in the issue template.

**Comment by ojw28** [comment]
Also, what branch is "exoplayer v2 branch as of today". There are two branches for v2 (dev and release).

**Comment by ojw28** [comment]
Figured it out; ignore me :). This is an interesting bug. We do correctly discard and re-buffer a different format if the track selector indicates that a different stream should be selected after preparation completes, and the decoder doesn't ever receive an input buffer in the discarded format.

There is however a separate optimization that's supposed to allow the format to be delivered to the decoder a little sooner, and this optimization is allowing the format of the discarded media to be delivered to the decoder.

We'll get that fixed. Note that once it's fixed, it's probably the case that you'll no longer need to do anything manual in your application code to restrict the streams that are being selected. The default track selection logic is supposed to do this automatically.

**Comment by ojw28** [comment]
As an aside for producing your streams: HLS players typically download a segment of the first variant listed in the master playlist at the start of playback. So you shouldn't put the highest quality variant first. [Apple's best practices](https://developer.apple.com/library/content/technotes/tn2224/_index.html#//apple_ref/doc/uid/DTS40009745-CH1-BITRATERECOMMENDATIONS) state:

> The first entry in the master playlist will be played at the initiation of a stream and is used as part of a test to determine which stream is most appropriate. The order of the other streams is irrelevant. Therefore, the first bit rate in the playlist should be the one that most clients can sustain.

**Comment by davibe** [comment]
You are right, I am sorry I was not very precise while addressing this. I have updated the description to specify that I am using `release-v2`.
This is the output of `adb bugreport` taken right after opening the media reproducing the issue.
[bugreport-1.txt](https://github.com/google/ExoPlayer/files/1156241/bugreport-1.txt)

Despite Apple guidelines the server generating that manifest is not directly under my control.

**Comment by davibe** [comment]
Finally. This is the test video to reproduce the issue (using Samsung S4 Mini) http://dmvideo.deltatre.it/VR/TEST_MBR/main.m3u

**Comment by ojw28** [comment]
- This should be fixed on the `dev-v2` branch now. Please verify. Not that you shouldn't need to do anything explicit in application code to restrict the formats being used to those that the device can play, since ExoPlayer's `DefaultTrackSelector` should be doing that for you already (unless you're overriding it).
- It's worth reporting the issue with the ordering in the master playlist to whoever is producing the content. Putting the highest quality variant first is against the guidelines and likely to cause issues for other players. It's also inefficient, since it means clients will download a chunk that's significantly larger than necessary during initialization. This may increase the time required for initial buffering.

**Comment by davibe** [comment]
I can confirm your fix in `dev-v2` works fine with the incriminated video.

Could you please tell me when this gets released in a version I can use directly from gradle ? I tried to make it work from the git branch (using jitpack) but it fails while building and I was not able to make it work.

**Comment by ojw28** [comment]
If you're happy to depend directly on the `dev-v2` branch then you can follow the "Locally" instructions in its [README](https://github.com/google/ExoPlayer/blob/dev-v2/README.md). Else you will need to wait for the 2.5 release. It's proving quite a tricky one to get finalized because it's so big. Hopefully in a couple of weeks.
